### PR TITLE
Use recently added allocation ids for shard started/failed

### PR DESCRIPTION
**Pull request by bleskes**
#12242 introduced the notion of allocation id - a unique identifier that is changed every time a shard is assigned to a node. This new id is a much utility to do the shard matching needed when processing shard started/failed messages. No need to match index uuid, node id and shard state. This change simplifies the matching logic to use these allocation ids.

 On top of that:
 1) A relocation target shards' allocation id is changed to include the allocation id of the source shard under relocatingId (similar to shard routing semantics)
 2) The logic around state change for finalize shard relocation is simplified - one simple start the target shard (we previously had unused logic around relocating state)

### CODE REVIEW

**Comment by bleskes** [comment]
@kimchy can you take a look?

File: core/src/main/java/org/elasticsearch/cluster/routing/allocation/AllocationService.java [hunk 1]
```diff
@@ -322,50 +321,54 @@ private boolean deassociateDeadNodes(RoutingAllocation allocation) {
...
+                if (shard.allocationId().getId().equals(startedShard.allocationId().getId())) {
+                    if (shard.active()) {
+                        logger.trace("{} shard is already started, ignoring (routing: {})", startedShard.shardId(), startedShard);
+                    } else {
+                        dirty = true;
+                        // override started shard with the latest copy. Capture it now , before starting the shard destroys it...
+                        startedShard = new ShardRouting(shard);
+                        routingNodes.started(shard);
+                        logger.trace("{} marked shard as started (routing: {})", startedShard.shardId(), startedShard);
                         break;
```
Comment (kimchy) [inline core/src/main/java/org/elasticsearch/cluster/routing/allocation/AllocationService.java]: should the break be outside of the if/else? we found  match

File: core/src/main/java/org/elasticsearch/cluster/routing/allocation/AllocationService.java [hunk 2]
```diff
@@ -322,50 +321,54 @@ private boolean deassociateDeadNodes(RoutingAllocation allocation) {
     private boolean applyStartedShards(RoutingNodes routingNodes, Iterable<? extends ShardRouting> startedShardEntries) {
         boolean dirty = false;
         // apply shards might be called several times with the same shard, ignore it
+        STARTED_SHARDS_LOOP:
```
Comment (kimchy) [inline core/src/main/java/org/elasticsearch/cluster/routing/allocation/AllocationService.java]: left over? it is not used

File: core/src/main/java/org/elasticsearch/cluster/routing/allocation/AllocationService.java [hunk 3]
```diff
@@ -384,127 +387,100 @@ private boolean applyFailedShard(RoutingAllocation allocation, ShardRouting fail
...
+
+        // replace incoming copy to make sure we work on the latest copy
+        failedShard = matchedNode.current();
+
+        if (addToIgnoreList) {
+            // make sure we ignore this shard on the relevant node
+            allocation.addIgnoreShardForNode(failedShard.shardId(), failedShard.currentNodeId());
+        }
+
+        // handle relocation target shards
```
Comment (kimchy) [inline core/src/main/java/org/elasticsearch/cluster/routing/allocation/AllocationService.java]: can we add a better javadoc here? and that if we find it? Especially the early return statement at the end of this if, that one is confusing

File: core/src/main/java/org/elasticsearch/cluster/routing/ShardRouting.java [hunk 4]
```diff
@@ -567,11 +575,16 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
...
+                .field("index", shardId().index().name())
+                .field("version", version);
+
         if (restoreSource() != null) {
             builder.field("restore_source");
             restoreSource().toXContent(builder, params);
         }
+        if (allocationId != null) {
+            allocationId.toXContent(builder, params);
+        }
```
Comment (kimchy) [inline core/src/main/java/org/elasticsearch/cluster/routing/ShardRouting.java]: why did you add it? I mean, it is very internal...., it will mean cluster state API will become so much more noisy

File: core/src/main/java/org/elasticsearch/cluster/routing/ShardRouting.java [hunk 5]
```diff
@@ -502,6 +503,9 @@ public boolean equals(Object o) {
         if (relocatingNodeId != null ? !relocatingNodeId.equals(that.relocatingNodeId) : that.relocatingNodeId != null) {
             return false;
         }
+        if (allocationId != null ? !allocationId.equals(that.allocationId) : that.allocationId != null) {
+            return false;
+        }
```
Comment (kimchy) [inline core/src/main/java/org/elasticsearch/cluster/routing/ShardRouting.java]: this change requires going over all the places we use `equals` in ShardRouting....

**Comment by kimchy** [comment]
@bleskes left some comments, it is so much cleaner now :)

File: core/src/main/java/org/elasticsearch/cluster/routing/ShardRouting.java (hunk 5 above)
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/cluster/routing/ShardRouting.java]: I went through all equals I could find and ShardRouting as a key type in map. Didn't spot only problem. The only shady place is IndicesClusterService, but that's up next for cleaning.  I also think it's not a big risk as the allocation id is changed only on state changes , so other fields in ShardRouting are changed as well.

File: core/src/main/java/org/elasticsearch/cluster/routing/ShardRouting.java (hunk 4 above)
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/cluster/routing/ShardRouting.java]: I was weighing the down sides against the upsides and ended up thinking that we end up pulling the entire cluster state with mapping and routing tables for debugging purposes and in that cases I rather have more info than miss something. I'm not sure it adds so much noise. If it helps I can inline as allocation_id and relocating_allocation_id. If you feel strongly about it I'll remove. I just have a gut feeling I'll miss it some day ;)

File: core/src/main/java/org/elasticsearch/cluster/routing/allocation/AllocationService.java (hunk 2 above)
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/cluster/routing/allocation/AllocationService.java]: yep. will removed.

File: core/src/main/java/org/elasticsearch/cluster/routing/allocation/AllocationService.java (hunk 1 above)
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/cluster/routing/allocation/AllocationService.java]: yeah.. will cahnged.

**Comment by bleskes** [comment]
@kimchy thx. Pushed another commit.

**Comment by kimchy** [comment]
LGTM

File: core/src/main/java/org/elasticsearch/cluster/routing/RoutingNodes.java [hunk 6]
```diff
@@ -785,10 +793,16 @@ public void remove() {
         }
 
         public void moveToUnassigned(UnassignedInfo unassignedInfo) {
-            remove();
+            if (isRemoved() == false) {
```
Comment (boliza) [inline core/src/main/java/org/elasticsearch/cluster/routing/RoutingNodes.java]: `if (!isRemoved())` is better ,is that right?
